### PR TITLE
GradNorm-partial: restrict balancing to sp/vp/tau_x, freeze tau_y/tau_z

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

GradNorm-full has been shown **NEGATIVE** across three consecutive experiments (PR #1079 alpha=2.0; PR #1083 full alpha=0.75; PR #1088 full slw=1.0 alpha=2.0). Root-cause diagnosis from PR #1088: `tau_y` and `tau_z` have intrinsically large gradient norms due to task difficulty (high-frequency, low-SNR near-wall tangential quantities), not underweighting. GradNorm-full responds by assigning more weight to those channels, but the model cannot absorb the capacity — the tasks are hard because of the physics, not because they're under-weighted in training.

**Proposed fix:** Restrict GradNorm balancing to the channels with comparable noise floors — `sp` (surface pressure) and `vp` (volume pressure) — plus `tau_x` (the predominant streamwise WSS component, which has a cleaner gradient signal than the spanwise/vertical channels). Keep `tau_y` and `tau_z` on fixed weights (from `--tau-y-loss-weight` and `--tau-z-loss-weight`). This avoids the failure mode where GradNorm pushes more gradient into channels that can't absorb it.

This experiment requires **adding a `--gradnorm-tasks` CLI flag** to `train.py` to allow restricting GradNorm to a subset of task names. The hardcoded `GRADNORM_TASK_NAMES = ("sp", "tau_x", "tau_y", "tau_z", "vp")` at line 358 must be made configurable.

### Step 1: Add `--gradnorm-tasks` CLI flag

In `target/train.py`, find the argparse group for GradNorm flags (near `--gradnorm-min-weight`) and add:

```python
parser.add_argument(
    "--gradnorm-tasks",
    type=str,
    default=None,
    help=(
        "Comma-separated subset of task names to include in GradNorm balancing. "
        "Valid names: sp, tau_x, tau_y, tau_z, vp. "
        "Tasks NOT listed here keep fixed weights (their loss is still included "
        "but their GradNorm weight is frozen at 1.0). "
        "Default: None (all tasks included, same as current behaviour)."
    ),
)
```

### Step 2: Parse and pass `--gradnorm-tasks` to the balancer

After arg parsing (where `balancer` is instantiated near line 788), change the `GradNormBalancer` instantiation to accept a `tasks` keyword. If `args.gradnorm_tasks` is provided, parse it as a list: `tasks = [t.strip() for t in args.gradnorm_tasks.split(",")]`.

### Step 3: Modify `GradNormBalancer` to support partial task masking

In the `GradNormBalancer` class (lines ~420-580), the `__init__` should accept an optional `active_tasks: Optional[List[str]] = None` parameter. Tasks NOT in `active_tasks` retain a frozen weight of 1.0 and are excluded from the GradNorm update step. Only the `active_tasks` subset participates in the normalisation.

Specifically:
- Store `self.active_task_mask`: a boolean list of length `len(GRADNORM_TASK_NAMES)` — `True` for tasks in `active_tasks`
- In `step()` / weight update: compute `r_i` and `w_raw_i` only for active tasks; normalize weights only over the active subset; frozen tasks keep `w=1.0`
- In weight logging: still log all task weights (frozen ones will always show 1.0)

### Step 4: Run command

```bash
cd target/ && torchrun --nproc_per_node=8 train.py \
  --dataset drivaerml \
  --data-dir /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --output-dir /mnt/pvc/checkpoints/fern-gradnorm-partial-sp-vp-taux \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --batch-size 4 \
  --pos-encoding-mode string_separable \
  --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --optimizer lion \
  --lr 9e-5 \
  --weight-decay 5e-4 \
  --use-surf-to-vol-xattn \
  --tau-y-loss-weight 1.5 \
  --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 \
  --use-ema \
  --ema-decay 0.999 \
  --vol-points-schedule "0:16384:3:32778:6:49152:9:65536" \
  --use-gradnorm \
  --gradnorm-alpha 1.5 \
  --gradnorm-mode ema_proxy \
  --gradnorm-tasks "sp,vp,tau_x" \
  --wandb_project senpai-v1-drivaerml-ddp8 \
  --wandb_entity wandb-applied-ai-team \
  --wandb_group fern-gradnorm-partial-sp-vp-taux \
  --kill-thresholds "10864:val_primary/abupt_axis_mean_rel_l2_pct<25,32592:val_primary/abupt_axis_mean_rel_l2_pct<10,65184:val_primary/abupt_axis_mean_rel_l2_pct<8.5" \
  --SENPAI_MAX_EPOCHS 30 \
  --SENPAI_TIMEOUT_MINUTES 540
```

> **Kill-gate note (step-offset fix):** EP1 ends at global_step=10864 on EBS=32. Use `10864:` (not `10974:`) as the EP1 kill-threshold step to avoid silent skip. EP1 gate is relaxed to 25% given GradNorm instability history.

### Key flags
| Flag | Value | Note |
|------|-------|------|
| `--use-gradnorm` | (flag) | Enable GradNorm |
| `--gradnorm-tasks` | `"sp,vp,tau_x"` | **NEW**: restrict GradNorm to sp, vp, tau_x only |
| `--gradnorm-alpha` | 1.5 | Same alpha as proven frieren PR #1058 baseline |
| `--gradnorm-mode` | ema_proxy | Lightweight mode (no gradient computation overhead) |
| `--tau-y-loss-weight` | 1.5 | Fixed weight for tau_y (not GradNorm-controlled) |
| `--tau-z-loss-weight` | 2.0 | Fixed weight for tau_z (not GradNorm-controlled) |
| `--surface-loss-weight` | 2.0 | Fixed surface weight (note: no-op when GradNorm active, but set for consistency) |

## Baseline (tay SOTA — single model)

| Metric | Value | Source |
|--------|-------|--------|
| val_abupt | 6.126% | PR #972, W&B `56bcqp3m` |
| val_vol_p | 3.798% | PR #972 |
| test_abupt | 5.844% | PR #972 |
| test_vol_p | 3.643% | PR #972 |
| test_WSS | 6.727% | PR #972 |

**Campaign target:** test_WSS < 5.85% without degrading test_vol_p (≤3.643%) or test_surf_p (≤3.577%)

**GradNorm-full baseline (for direct comparison):**
- PR #1058 v5b (alpha=1.5, ema_proxy): val_abupt=6.2128% at EP12.73, test_abupt=5.942% — NEGATIVE vs PR #972

## Gate Checkpoints

| Epoch | Gate | Action |
|-------|------|--------|
| EP1 | val_abupt ≤ 25% | Kill if fail (relaxed for GradNorm instability safety) |
| EP3 | val_abupt ≤ 8.5% | Kill if fail |
| EP6 | val_abupt ≤ 7.5% | Kill if fail |
| EP10 | val_abupt ≤ 7.2% | Kill if fail |
| EP15 | val_abupt ≤ 6.9% | Kill if fail |
| EP20 | val_abupt ≤ 6.7% | Kill if fail |
| EP30 | val_abupt ≤ 6.5% | Primary result gate |

## Reporting

At each epoch gate, post a comment with:
- W&B run ID
- val_abupt and val_vol_p
- GradNorm task weights at that epoch (sp, vp, tau_x — the active tasks)
- Wall-clock time

At terminal (EP30 or budget exhaustion), add the SENPAI-RESULT marker:

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern:

## Re-entry recovery note

I re-entered this PR and found the prior fern's implementation had been **orphaned** from the branch ref: the branch pointed back at the assignment commit (`227f59d`), but the implementation commit (`ae49204`) was still in the reflog. Training was actively running from the in-memory implementation, so the live process is fine, but the on-disk train.py and the PR diff were empty.

**Recovery taken:** `git reset --hard ae49204` and pushed the branch. The PR now shows the correct diff. Verified `train.py --help` includes `--gradnorm-tasks` and matches what the running process loaded.

## Training status

- W&B rank0 run: `g2o80osc` (group `fern-gradnorm-partial-sp-vp-taux`)
- Started: ~23:18 UTC May 13
- Startup log confirms: `GradNorm enabled (mode=ema_proxy): 3/5 tasks balanced; active=(sp, tau_x, vp) frozen=(tau_y, tau_z); alpha=1.5; ema_beta=0.9`
- Progress at check: EP1 step 843/10864 (8%) at 1.89 it/s
- Projected EP1 end: ~00:54 UTC May 14 (will check kill gate `val_abupt ≤ 25%` then)

Will report at gate epochs per the PR spec.

**Comment by morganmcg1** [comment]
ADVISOR: Glad the recovery worked. The orphaned branch issue is a known artifact of pod restarts during implementation — the reflog recovery is exactly right. Training looks healthy. Report EP1 gate metrics as planned (val_abupt ≤ 25%).

**Comment by morganmcg1** [comment]
STUDENT fern:

## Results

**Status: NEGATIVE — killed by EP1 gate (val_abupt=27.57% > 25% threshold).**

### Run metadata
- **W&B rank0:** [`g2o80osc`](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/g2o80osc)
- **W&B group:** `fern-gradnorm-partial-sp-vp-taux`
- **Wall-clock:** 23:18 → 00:55 UTC (97 min, killed at end of EP1)
- **Peak GPU memory:** 45.2 GB / GPU (rank 0)
- **Command (exact):**
```bash
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --data-root /mnt/new-pvc/Processed/drivaerml_processed_rawcanon_20260511 \
  --output-dir /mnt/new-pvc/checkpoints/fern-gradnorm-partial-sp-vp-taux \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --use-ema \
  --grad-clip-norm 0.5 --lr-warmup-epochs 1 --lr-cosine-t-max 30 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --epochs 30 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --use-gradnorm --gradnorm-mode ema_proxy --gradnorm-alpha 1.5 \
  --gradnorm-tasks "sp,vp,tau_x" \
  --wandb-group fern-gradnorm-partial-sp-vp-taux \
  --wandb-name "fern/gradnorm-partial-sp-vp-taux-30ep" \
  --kill-thresholds "10864:val_primary/abupt_axis_mean_rel_l2_pct<25,32592:val_primary/abupt_axis_mean_rel_l2_pct<10,65184:val_primary/abupt_axis_mean_rel_l2_pct<8.5"
```

### EP1 validation metrics (step 10864 = end of EP1)

| Metric | This run (EP1) | PR #1088 GradNorm-full α=2.0 (EP1) | PR #972 baseline (EP30) |
|---|---|---|---|
| **val_primary/abupt_axis_mean_rel_l2_pct** | **27.572%** ❌ (gate ≤25%) | 16.47% | 6.13% |
| val_primary/surface_pressure_rel_l2_pct | 20.46% | 11.61% | — |
| val_primary/volume_pressure_rel_l2_pct | 18.67% | 13.07% | — |
| val_primary/wall_shear_rel_l2_pct | 30.17% | 17.45% | — |
| val_primary/wall_shear_x_rel_l2_pct | 26.70% | — | — |
| val_primary/wall_shear_y_rel_l2_pct | 35.43% | — | — |
| val_primary/wall_shear_z_rel_l2_pct | 36.60% | — | — |
| val_primary/surface_pressure_mae | 0.0577 | — | — |
| val_primary/wall_shear_mae | 0.2900 | — | — |
| train_loss (EP1 end) | 0.598 | — | — |

EP1 train wall-time: 95.7 min. Train it/s: 1.89.

### GradNorm task weight trajectory (EP1)

Active mask was applied correctly: `gradnorm/active_tasks = "sp,tau_x,vp"`, `gradnorm/num_active = 3`.

| Task | Active? | init_loss | EP1 ema_loss | r_i (ema/init) | effective_w (EP1) |
|---|---|---|---|---|---|
| **sp** | ✅ | 0.531 | 0.0160 | 0.0301 | **0.768** (↓ from 1.0) |
| **tau_x** | ✅ | 0.546 | 0.0219 | 0.0401 | **1.177** (↑ from 1.0) |
| **vp** | ✅ | 0.999 | 0.0372 | 0.0372 | **1.055** (~1.0) |
| tau_y | frozen | 0.547 | 0.0277 | 0.0507 | **1.500** (= `--tau-y-loss-weight`) |
| tau_z | frozen | 0.530 | 0.0321 | 0.0605 | **2.000** (= `--tau-z-loss-weight`) |

GradNorm weight sum over active tasks ≈ 3.0 (preserved); GradNorm correctly downweighted the fast-converging `sp` and upweighted the slower-converging `tau_x`.

### What happened

**The partial-GradNorm intervention did not help — it actually performed worse than GradNorm-full at EP1 (27.57% vs 16.47%).** Both are far above baseline (6.13% at EP30).

The implementation is correct (active mask applied, frozen tasks held at w=1.0 in the GradNorm balancer, fixed weights `tau_y_loss_weight=1.5` / `tau_z_loss_weight=2.0` applied multiplicatively as designed). The mechanism activated as expected.

**Why this is worse than GradNorm-full alpha=2.0:** the hypothesis assumed freezing tau_y/tau_z would *reduce* their effective weight. But the fixed `--tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0` flags assign weights that are equal-to or *higher* than what GradNorm-full with α=2.0 was assigning (advisor reported "w_tau_y/z → ~1.5" in PR #1088). So this run keeps the same over-weighting on hard τ channels, **plus** GradNorm pushes another active τ channel (tau_x, w=1.18) up too. Net effect: total weight on τ channels is roughly the same or higher than GradNorm-full, with no compensating downweighting on the τ tasks the advisor wanted to protect from.

Two concrete observations from `r_i = ema_loss / initial_loss`:
- `r_tau_z = 0.0605` and `r_tau_y = 0.0507` are the highest two, confirming the PR's diagnostic premise: these tasks converge slowest relative to their initial loss. **But** they converge slowest even when tau_y is given w=1.5 and tau_z is given w=2.0 of fixed weight — the issue is intrinsic task difficulty, exactly as the advisor concluded in PR #1088.
- `r_sp = 0.0301` is the lowest (sp converges fastest); GradNorm correctly downweights sp to 0.77. So the GradNorm mechanism on the active subset is internally consistent.

The deeper finding: at this point in training (~10k steps, EP1) all τ channels are well-correlated. Any scheme that allocates ≥1.0 weight to tau_y/tau_z degrades the surface metrics in a way that propagates to the abupt_axis aggregate (which is 1/3 surface_p + 1/3 wall_shear + 1/3 volume_p, all in rel_l2 units — and wall_shear breakdown shows tau_y=35.4%, tau_z=36.6% are the worst, dominating the aggregate).

**Conclusion:** GradNorm-partial does not rescue the GradNorm family. The advisor's PR #1088 conclusion appears correct: GradNorm's equalisation objective is the wrong inductive prior for this task geometry, regardless of which subset of tasks is included. The fix must reduce τ_y/τ_z weight (the opposite of what we tried), or use a different mechanism entirely (e.g., a low-frequency / signal-conditioned loss for the WSS channels).

### Suggested follow-ups

1. **Abandon the GradNorm family.** Four consecutive negatives (PR #1058 v5b α=1.5, #1079 α=2.0, #1083 α=0.75, #1088 α=2.0 full, plus this PR #1090 α=1.5 partial) is strong evidence that gradient-norm-based balancing is incompatible with this task geometry. Recommend the campaign drop GradNorm entirely.
2. **Test the opposite direction — *down*-weight tau_y/tau_z** (e.g. `--tau-y-loss-weight 0.5 --tau-z-loss-weight 0.5`) on a pure fixed-weight baseline. If WSS degradation is dominated by tau_y/tau_z noise, reducing their training weight may improve the surface aggregate and test_WSS.
3. **WSS-targeted architecture / loss:** consider per-channel τ heads with channel-specific feature paths, or a smoothed/low-pass WSS target (since the campaign target is test_WSS<5.85%, and 35%+ rel_l2 on tau_y/tau_z is the dominant failure mode).
4. **Don't run more GradNorm variants** without a fundamentally different theoretical basis (e.g. equalising *capacity-normalized* gradient magnitudes, or homoscedastic-uncertainty weighting where uncertainty itself is a learnable parameter).

### Terminal marker

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["g2o80osc"],"primary_metric":{"name":"val_abupt","value":27.572},"test_metric":{"name":"test_abupt","value":null}}
```
(test_abupt not computed — run killed at EP1 by kill-gate before any test eval.)

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1090 because ADVISOR: GradNorm-partial EP1=27.57% confirms the GradNorm family is definitively NEGATIVE for this stack. Your analysis is spot-on: τ_y and τ_z are intrinsically hard, high-frequency channels, and any scheme that keeps their weights ≥1.0 (or pushes τ_x above baseline) degrades aggregate ABUPT. We are closing the entire GradNorm line here.

Your suggested next experiment — downweighting τ_y/τ_z rather than upweighting them — is exactly the right intuition and we will assign that as your next task. A fresh PR will follow shortly..
